---
title: Processing the Grun pancreas dataset
author: Aaron Lun
---

```{r style, echo=FALSE, results='hide', message=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Downloading the count data

We obtain a single-cell RNA sequencing dataset of human pancreas from [Grun et al. (2016)](https://doi.org/10.1016/j.stem.2016.05.010).
A count matrix for endogenous genes and spike-ins is provided from the Gene Expression Omnibus 
using accession code [GSE81076](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81076).
We download it using `r Biocpkg("BiocFileCache")` to cache the results:

```{r}
library(BiocFileCache)
bfc <- BiocFileCache("raw_data", ask=FALSE)
grun.fname <- bfcrpath(bfc, paste0("ftp://ftp.ncbi.nlm.nih.gov/geo/series/",
    "GSE81nnn/GSE81076/suppl/GSE81076%5FD2%5F3%5F7%5F10%5F17%2Etxt%2Egz"))
```

We read the table into memory as a sparse matrix.

```{r}
library(scuttle)
counts <- readSparseCounts(grun.fname, quote="\"")
dim(counts)
```

# Forming a `SingleCellExperiment`

```{r}
library(SingleCellExperiment)
library(SparseArray)
dn <- dimnames(counts)
dimnames(counts) <- NULL
y <- as(counts, "SVT_SparseMatrix") # don't convert to integer, these are numeric!
sce <- SingleCellExperiment(list(counts=y))
dimnames(sce) <- dn
```

Fleshing out the column metadata:

```{r}
lib.names <- sub("_.*", "", colnames(sce))
donor.names <- sub("(D10|D17|D2|D3|D7).*", "\\1", lib.names)
treatment <- c(
    D10631="CD63+ sorted cells",
    D101="live sorted cells, library 1",
    D102="live sorted cells, library 2",
    D1713="CD13+ sorted cells",
    D172444="CD24+ CD44+ live sorted cells",
    D17All1="live sorted cells, library 1",
    D17All2="live sorted cells, library 2",
    D17TGFB="TGFBR3+ sorted cells",
    D2ex="exocrine fraction, live sorted cells",
    D3en1="live sorted cells, library 1",
    D3en2="live sorted cells, library 2",
    D3en3="live sorted cells, library 3",
    D3en4="live sorted cells, library 4",
    D3ex="exocrine fraction, live sorted cells",
    D71="live sorted cells, library 1",
    D72="live sorted cells, library 2",
    D73="live sorted cells, library 3",
    D74="live sorted cells, library 4"
)[lib.names]
colData(sce) <- DataFrame(donor=donor.names, sample=treatment, row.names=colnames(sce))
```

Splitting up gene information:

```{r}
symbol <- sub("__.*", "", rownames(sce))
loc <- sub(".*__", "", rownames(sce))
rowData(sce) <- DataFrame(symbol=symbol, chr=loc)
```

Splitting spike-ins into an alternative experiment:

```{r}
status <- ifelse(grepl("ERCC-[0-9]+", symbol), "ERCC", "endogenous")
sce <- splitAltExps(sce, status, ref="endogenous")

spike.exp <- altExp(sce, "ERCC")
spikedata <- scRNAseq::ERCCSpikeInConcentrations(volume = 20, dilution = 50000)
spikedata <- spikedata[rownames(spike.exp), ]
rowData(spike.exp) <- cbind(rowData(spike.exp), spikedata)
altExp(sce, "ERCC") <- spike.exp
```

# Saving to file

We save all of the components to file for upload to `r Biocpkg("ExperimentHub")`.

```{r}
library(scRNAseq)
unlink('finished', recursive=TRUE)
saveDataset(sce, "finished",
    createMetadata(
        title="De Novo Prediction of Stem Cell Identity using Single-Cell Transcriptome Data",
        description="Adult mitotic tissues like the intestine, skin, and blood undergo constant turnover throughout the life of an organism. Knowing the identity of the stem cell is crucial to understanding tissue homeostasis and its aberrations upon disease. Here we present a computational method for the derivation of a lineage tree from single-cell transcriptome data. By exploiting the tree topology and the transcriptome composition, we establish StemID, an algorithm for identifying stem cells among all detectable cell types within a population. We demonstrate that StemID recovers two known adult stem cell populations, Lgr5+ cells in the small intestine and hematopoietic stem cells in the bone marrow. We apply StemID to predict candidate multipotent cell populations in the human pancreas, a tissue with largely uncharacterized turnover dynamics. We hope that StemID will accelerate the search for novel stem cells by providing concrete markers for biological follow-up and validation.

This dataset is represented by a `SingleCellExperiment` containing the pancreas cells from the study. The sorting protocol and donor identity for each cell is obtained from the sample annotations in the GEO entry. ERCC spike-ins are available in the alternative experiment.",
        taxonomy.id="9606",
        genome="GRCh37",
        sources=list(list(provider="GEO", id='GSE81076')),
        maintainer.name="Aaron Lun",
        maintainer.email="infinite.monkeys.with.keyboards@gmail.com"
    )
)
```

# Session information

```{r}
sessionInfo()
```
