---
title: "Overview of the scRNAseq dataset collection"
author: "Davide Risso"
date: "Created: May 25, 2016; Compiled: `r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
package: scRNAseq
vignette: >
  %\VignetteIndexEntry{User's Guide}
  %\VignetteEngine{knitr::rmarkdown}
bibliography: biblio.bib
---

```{r style, echo=FALSE}
knitr::opts_chunk$set(error=FALSE, warning=FALSE, message=FALSE)
```

# Introduction

The `r Biocpkg("scRNAseq")` package provides convenient access to several publicly available data sets 
in the form of `SingleCellExperiment` objects.
The focus of this package is to capture datasets that are not easily read into R with a one-liner from, e.g., `read.csv()`.
Instead, we do the necessary data munging so that users only need to call a single function to obtain a well-formed `SingleCellExperiment`.
For example:

```{r}
library(scRNAseq)
fluidigm <- ReprocessedFluidigmData()
fluidigm
```

Readers are referred to the `r Biocpkg("SummarizedExperiment")` and `r Biocpkg("SingleCellExperiment")` documentation 
for further information on how to work with `SingleCellExperiment` objects.

# Available data sets

The available data sets can be split into two categories.
The first category contains expression matrices that have been generated by the `r Biocpkg("scRNAseq")` authors 
from the raw sequencing data for each experiment.
This includes:

- `ReprocessedFluidigmData()` provides 65 cells from @pollen2014low.
- `ReprocessedTh2Data()` provides 96 T helper cells from @mahata2014single.
- `ReprocessedAllenData()` provides 379 cells from the mouse visual cortex,
   which is a subset of the data from @tasic2016adult.

The second category contains expression matrices that were provided by the authors of each study.
No further reprocessing has been performed other than some cross-checks betweeh the count matrix and the sample metadata.

- `TasicBrainData()` provides 1809 cells (and some control libraries) from the mouse brain [@tasic2016adult].
- `ZeiselBrainData()` provides 3005 cells from the mouse brain [@zeisel2015brain].
- `SegerstolpePancreasData()` provides 3514 cells from human pancreas [@segerstolpe2016single].

# Adding new data sets

Please contact us if you have a data set that you would like to see added to this package.
The only requirement is that your data set has publicly available expression values (ideally counts) and sample annotation.
The more difficult/custom the format, the better, 
as its inclusion in this package will provide more value for other users in the R/Bioconductor community.

If you have already written code that processes your desired data set in a `SingleCellExperiment`-like form,
we would welcome a pull request [here](https://github.com/LTLA/scRNAseq).
The process can be expedited by ensuring that you have the following files:

- `inst/scripts/make-X-Y-data.Rmd`, a Rmarkdown report that creates all components of a `SingleCellExperiment`.
`X` should be the last name of the first author of the relevant study while `Y` should be the name of the biological system.
- `inst/scripts/make-X-Y-metadata.R`, an R script that creates a metadata CSV file at `inst/extdata/metadata-X-Y.csv`.
Metadata files should follow the format described in the `r Biocpkg("ExperimentHub")` documentation.
- `R/XYData.R`, an R source file that defines a function `XYData()` to download the components from ExperimentHub
and creates a `SingleCellExperiment` object.

Potential contributors are recommended to examine some of the existing scripts in the package to pick up the coding conventions.
Remember, we're more likely to accept a contribution if it's indistinguishable from something we might have written ourselves!

As a general rule, 10X Genomics data sets are not suitable for inclusion in this package.
They are either easy to load (e.g., with functions from the `r Biocpkg("DropletUtils")` package),
or they are more appropriately obtained with dedicated 10X packages like `r Biocpkg("TENxPBMCData")` or `r Biocpkg("TENxBrainData")`.
That said, inclusion will be considered if the format has been sufficiently customized by the original authors.

# References
