---
title: Processing the Usoskin brain dataset
author: Aaron Lun
date: "June 8, 2019"
output:
  BiocStyle::html_document:
    titlecaps: false
    toc_float: true
bibliography: ref.bib
---

```{r style, echo=FALSE, results='hide', message=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Overview

The @usoskin2015unbiased study provides a single-cell RNA seq dataset involving the mouse brain. 
This contains approximately 700 cells of varying types such as oligodendrocytes, microglia and neurons.
Libraries were prepared from individual cells using the STRT-Seq protocol.
After sequencing, expression was quantified by counting the number of reads mapped to each gene.

# Downloading the count data

Counts for endogenous genes and repeat regions are available as Supplementary Table 2 at http://linnarssonlab.org/drg/.
We download and cache them using the `r Biocpkg("BiocFileCache")` package.

```{r}
library(BiocFileCache)
bfc <- BiocFileCache("raw_data", ask = FALSE)
base.url <- file.path("https://storage.googleapis.com",
    "linnarsson-lab-www-blobs/blobs/drg")
count.file <- bfcrpath(bfc, file.path(base.url,
    "Usoskin%20et%20al.%20External%20resources%20Table%202.xlsm"))
```

We load them into memory.

```{r}
library(readxl)
all.raw <- read_xlsx(count.file, sheet=6)
all.raw <- all.raw[,1:706] # last two columns are empty!
```

# Unpacking the format

We need to unpack the column data:

```{r}
meta.raw <- head(all.raw, 9)
coldata <- data.frame(t(meta.raw[,-(1:10)]), stringsAsFactors=FALSE)
colnames(coldata) <- meta.raw[,10,drop=TRUE]
head(coldata)
```

And then the rowdata:

```{r}
gene.info <- tail(all.raw, -12)
rowdata <- as.data.frame(gene.info[,1:10])
rowdata.names <- all.raw[11,1:10,drop=TRUE]
rowdata.names <- sub("^[0-9]\\. ", "", rowdata.names)
colnames(rowdata) <- rowdata.names
dim(rowdata)
colnames(rowdata)
```

And finally, the RPMs (since no raw counts have been provided):

```{r}
rpms <- as.matrix(gene.info[,-(1:10)])
storage.mode(rpms) <- "numeric"
dim(rpms)
```

We move gene and sample IDs into the row and column names.

```{r}
rownames(rpms) <- rowdata[,1]
colnames(rpms) <- coldata[,1]
rowdata <- rowdata[,-1]
coldata <- coldata[,-1]
```

# Saving to file 

We now save all of the components to file for upload to `r Biocpkg("ExperimentHub")`.
These will be used to construct a `SingleCellExperiment` on the client side when the dataset is requested.

```{r}
path <- file.path("scRNAseq", "usoskin-brain", "2.0.0")
dir.create(path, showWarnings=FALSE, recursive=TRUE)
saveRDS(rpms, file=file.path(path, "rpms.rds"))
saveRDS(coldata, file=file.path(path, "coldata.rds"))
saveRDS(rowdata, file=file.path(path, "rowdata.rds"))
```

# Session information

```{r}
sessionInfo()
```

# References
