---
title: "Processing Kotliarov-2020 CITEseq data (used in dsb paper)"
author: "Stephany Orjuela"
---

```{r setup, echo=FALSE, results='hide', message=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Download the data

We obtain a single-cell RNA sequencing dataset of the human PBMCs from [Kotliarov et al. (2020)](https://doi.org/10.1038/s41591-020-0769-8).
Counts for endogenous genes and antibody-derived tags (ADTs) are available from [figshare](https://nih.figshare.com/collections/Data_and_software_code_repository_for_Broad_immune_activation_underlies_shared_set_point_signatures_for_vaccine_responsiveness_in_healthy_individuals_and_disease_activity_in_patients_with_lupus_Kotliarov_Y_Sparks_R_et_al_Nat_Med_DOI_https_d/4753772).

Code used to analyze the data is available in the same link.

Of particular interest is the `H1_day0_demultilexed_singlets.RDS` file, 
which is a Seurat object that contains demultiplexed counts (filtered for singlets). 
Demultiplexing was performed with HTO counts and verified with demuxlet. 
Counts available have not been filtered by QC or normalized.

```{r}
library(BiocFileCache)
bfc <- BiocFileCache("raw_data", ask = FALSE)
url <- "https://nih.figshare.com/ndownloader/files/20706642"
contents <- bfcrpath(bfc, url)
seuratObj <- readRDS(contents)
```

# Extract elements from Seurat object

First we extract the RNA counts:

```{r}
library(S4Vectors)
rna.mat <- seuratObj@data
dim(rna.mat)

coldata <- DataFrame(seuratObj@meta.data)
nrow(coldata)
colnames(coldata)
```

Then the ADT counts:

```{r}
adt.mat <- seuratObj@assay$CITE@raw.data
dim(adt.mat)
```

# Assembly into a `SingleCellExperiment`
 
We assemble this into a `SingleCellExperiment`:

```{r}
library(SingleCellExperiment)
sce <- SingleCellExperiment(list(counts=rna.mat), colData=coldata)
```

Adding the ADTs, with some clean-ups to the row data:

```{r}
se <- SummarizedExperiment(list(counts=adt.mat))

targets <- sub("_.*$", "", rownames(se))
targets <- gsub(" ", "", targets)
targets <- sub("k(:appa)?isotype$", "κ", targets, ignore.case=TRUE)
rowData(se)$target <- targets
targets

rowData(se)$isotype <- grepl("isotype", rownames(se), ignore.case=TRUE)
altExp(sce, "ADT") <- se
rowData(se)
```

# Save for upload

We perform some polishing to optimize for space:

```{r}
library(scRNAseq)
sce <- polishDataset(sce)
sce
```

And we save it with some metadata:

```{r}
meta <- list(
    title="Broad immune activation underlies shared set point signatures for vaccine responsiveness in healthy individuals and disease activity in lupus patients",
    description="Responses to vaccination and to diseases vary widely across individuals, which may be partly due to baseline immune variations. Identifying such baseline predictors of immune responses and their biological basis are of broad interest given their potential importance for cancer immunotherapy, disease outcomes, vaccination and infection responses. Here we uncover baseline blood transcriptional signatures predictive of antibody responses to both influenza and yellow fever vaccinations in healthy subjects. These same signatures evaluated at clinical quiescence are correlated with disease activity in systemic lupus erythematosus patients with plasmablast-associated flares. CITE-seq profiling of 82 surface proteins and transcriptomes of 53,201 single cells from healthy high and low influenza-vaccination responders revealed that our signatures reflect the extent of activation in a plasmacytoid dendritic cell—Type I IFN—T/B lymphocyte network. Our findings raise the prospect that modulating such immune baseline states may improve vaccine responsiveness and mitigate undesirable autoimmune disease activities.",
    taxonomy_id="9606",
    genome="GRCh37",
    sources=list(
        list(provider="PubMed", id="32094927"),
        list(provider="URL", id=url, version=as.character(Sys.Date()))
    ),
    maintainer_name="Stephany Orjuela",
    maintainer_email="sorjuelal@gmail.com"
)

saveDataset(sce, "2023-12-20_output", meta)
```

# Session information {-}

```{r}
sessionInfo()
```
